src/AWS/Lambda/S3Event.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE DeriveGeneric     #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards   #-}
+
+module AWS.Lambda.S3Event
+  (
+    S3Event(..)
+  ) where
+
+import Data.Aeson
+import Data.HashMap.Strict
+import Data.Text (Text)
+import GHC.Generics
+
+data S3 a =
+  S3 {
+    s3SchemaVersion :: Text,
+    configurationId :: Text,
+    bucket          :: Bucket a,
+    s3Object        :: S3Object a
```

Comment:
I think object can be any `FromJSON` and that `S3Object` is one of a few possibilities. See https://docs.aws.amazon.com/lambda/latest/dg/eventsources.html#eventsources-s3-delete

Reply:
Maybe would be best to enumerate them from a sum type?

Reply:
Sure, that works provided they are enumerable. I don't really know where to look to see what all could be in this attribute. I guess you could just deploy something with `Value` and perform the various operations to see what events are possible?

Reply:
Yeah let's leave it for now, I hope that AWS releases a full spec one day (maybe there is one, but I couldn't find it...)

Reply:
Yeah we were only saying earlier how frustrating it is that AWS documentation doesn't provide any clear specs 😠 

So from my understanding event version 2.1 is the latest version. In the docs they say 2.0 is no longer valid (again another gripe with AWS is that the link Louis posted has `/latest` in the path but contains outdated examples). The closest thing I could find to a spec was here:

```
{  
   "Records":[  
      {  
         "eventVersion":"2.1",
         "eventSource":"aws:s3",
         "awsRegion":"us-west-2",
         "eventTime":The time, in ISO-8601 format, for example, 1970-01-01T00:00:00.000Z, when Amazon S3 finished processing the request,
         "eventName":"event-type",
         "userIdentity":{  
            "principalId":"Amazon-customer-ID-of-the-user-who-caused-the-event"
         },
         "requestParameters":{  
            "sourceIPAddress":"ip-address-where-request-came-from"
         },
         "responseElements":{  
            "x-amz-request-id":"Amazon S3 generated request ID",
            "x-amz-id-2":"Amazon S3 host that processed the request"
         },
         "s3":{  
            "s3SchemaVersion":"1.0",
            "configurationId":"ID found in the bucket notification configuration",
            "bucket":{  
               "name":"bucket-name",
               "ownerIdentity":{  
                  "principalId":"Amazon-customer-ID-of-the-bucket-owner"
               },
               "arn":"bucket-ARN"
            },
            "object":{  
               "key":"object-key",
               "size":object-size,
               "eTag":"object eTag",
               "versionId":"object version if bucket is versioning-enabled, otherwise null",
               "sequencer": "a string representation of a hexadecimal value used to determine event sequence, 
                   only used with PUTs and DELETEs"
            }
         },
         "glacierEventData": {
            "restoreEventData": {
               "lifecycleRestorationExpiryTime": "The time, in ISO-8601 format, for example, 1970-01-01T00:00:00.000Z, of Restore Expiry",
               "lifecycleRestoreStorageClass": "Source storage class for restore"
            }
         }
      }
   ]
}```

I wasn't sure whether I should create a sum type for the different S3Objects, or whether just having a maybe on the fields that are nullable was ok?

Reply:
Yeah!`Maybe` is a good idea 👍 I assume they at least keep backward compatibility between v2.0 and v2.1 so if we match v2.0 we should be safe for v1 too?

Reply:
I can't believe they don't publish schemas... The `Maybe` fields should work, and we'll know pretty quick if they don't. If you want to be extra careful, you can write an example function and try it out.